// ts-sql-query/TypeAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceTypeCast = exports.CustomBooleanTypeAdapter = void 0;
class CustomBooleanTypeAdapter {
    constructor(trueValue, falseValue) {
        this.trueValue = trueValue;
        this.falseValue = falseValue;
    }
    transformValueFromDB(value, type, next) {
        return next.transformValueFromDB(value, type);
    }
    transformValueToDB(value, type, next) {
        return next.transformValueToDB(value, type);
    }
}
exports.CustomBooleanTypeAdapter = CustomBooleanTypeAdapter;
class ForceTypeCast {
    transformValueFromDB(value, type, next) {
        return next.transformValueFromDB(value, type);
    }
    transformValueToDB(value, type, next) {
        return next.transformValueToDB(value, type);
    }
    transformPlaceholder(placeholder, type, _forceTypeCast, valueSentToDB, next) {
        return next.transformPlaceholder(placeholder, type, true, valueSentToDB);
    }
}
exports.ForceTypeCast = ForceTypeCast;
