// ts-sql-query/queryRunners/ChainedQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainedQueryRunner = void 0;
class ChainedQueryRunner {
    get database() {
        return this.queryRunner.database;
    }
    constructor(queryRunner) {
        this.queryRunner = queryRunner;
    }
    useDatabase(database) {
        return this.queryRunner.useDatabase(database);
    }
    getNativeRunner() {
        return this.queryRunner.getNativeRunner();
    }
    getCurrentNativeTransaction() {
        return this.queryRunner.getCurrentNativeTransaction();
    }
    execute(fn) {
        return this.queryRunner.execute(fn);
    }
    executeSelectOneRow(query, params = []) {
        return this.queryRunner.executeSelectOneRow(query, params);
    }
    executeSelectManyRows(query, params = []) {
        return this.queryRunner.executeSelectManyRows(query, params);
    }
    executeSelectOneColumnOneRow(query, params = []) {
        return this.queryRunner.executeSelectOneColumnOneRow(query, params);
    }
    executeSelectOneColumnManyRows(query, params = []) {
        return this.queryRunner.executeSelectOneColumnManyRows(query, params);
    }
    executeInsert(query, params = []) {
        return this.queryRunner.executeInsert(query, params);
    }
    executeInsertReturningLastInsertedId(query, params = []) {
        return this.queryRunner.executeInsertReturningLastInsertedId(query, params);
    }
    executeInsertReturningMultipleLastInsertedId(query, params = []) {
        return this.queryRunner.executeInsertReturningMultipleLastInsertedId(query, params);
    }
    executeInsertReturningOneRow(query, params = []) {
        return this.queryRunner.executeInsertReturningOneRow(query, params);
    }
    executeInsertReturningManyRows(query, params = []) {
        return this.queryRunner.executeInsertReturningManyRows(query, params);
    }
    executeInsertReturningOneColumnOneRow(query, params = []) {
        return this.queryRunner.executeInsertReturningOneColumnOneRow(query, params);
    }
    executeInsertReturningOneColumnManyRows(query, params = []) {
        return this.queryRunner.executeInsertReturningOneColumnManyRows(query, params);
    }
    executeUpdate(query, params = []) {
        return this.queryRunner.executeUpdate(query, params);
    }
    executeUpdateReturningOneRow(query, params = []) {
        return this.queryRunner.executeUpdateReturningOneRow(query, params);
    }
    executeUpdateReturningManyRows(query, params = []) {
        return this.queryRunner.executeUpdateReturningManyRows(query, params);
    }
    executeUpdateReturningOneColumnOneRow(query, params = []) {
        return this.queryRunner.executeUpdateReturningOneColumnOneRow(query, params);
    }
    executeUpdateReturningOneColumnManyRows(query, params = []) {
        return this.queryRunner.executeUpdateReturningOneColumnManyRows(query, params);
    }
    executeDelete(query, params = []) {
        return this.queryRunner.executeDelete(query, params);
    }
    executeDeleteReturningOneRow(query, params = []) {
        return this.queryRunner.executeDeleteReturningOneRow(query, params);
    }
    executeDeleteReturningManyRows(query, params = []) {
        return this.queryRunner.executeDeleteReturningManyRows(query, params);
    }
    executeDeleteReturningOneColumnOneRow(query, params = []) {
        return this.queryRunner.executeDeleteReturningOneColumnOneRow(query, params);
    }
    executeDeleteReturningOneColumnManyRows(query, params = []) {
        return this.queryRunner.executeDeleteReturningOneColumnManyRows(query, params);
    }
    executeProcedure(query, params = []) {
        return this.queryRunner.executeProcedure(query, params);
    }
    executeFunction(query, params = []) {
        return this.queryRunner.executeFunction(query, params);
    }
    executeBeginTransaction() {
        return this.queryRunner.executeBeginTransaction();
    }
    executeCommit() {
        return this.queryRunner.executeCommit();
    }
    executeRollback() {
        return this.queryRunner.executeRollback();
    }
    isTransactionActive() {
        return this.queryRunner.isTransactionActive();
    }
    executeInTransaction(fn, outermostQueryRunner) {
        return this.queryRunner.executeInTransaction(fn, outermostQueryRunner);
    }
    executeDatabaseSchemaModification(query, params = []) {
        return this.queryRunner.executeDatabaseSchemaModification(query, params);
    }
    executeConnectionConfiguration(query, params = []) {
        return this.queryRunner.executeConnectionConfiguration(query, params);
    }
    addParam(params, value) {
        return this.queryRunner.addParam(params, value);
    }
    addOutParam(params, name) {
        return this.queryRunner.addOutParam(params, name);
    }
    createResolvedPromise(result) {
        return this.queryRunner.createResolvedPromise(result);
    }
    executeCombined(fn1, fn2) {
        return this.queryRunner.executeCombined(fn1, fn2);
    }
    isMocked() {
        return this.queryRunner.isMocked();
    }
    lowLevelTransactionManagementSupported() {
        return this.queryRunner.isMocked();
    }
}
exports.ChainedQueryRunner = ChainedQueryRunner;
