// ts-sql-query/queryRunners/MariaDBPoolQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MariaDBPoolQueryRunner = void 0;
const PromiseBasedPoolQueryRunner_1 = require("./PromiseBasedPoolQueryRunner");
const MariaDBQueryRunner_1 = require("./MariaDBQueryRunner");
class MariaDBPoolQueryRunner extends PromiseBasedPoolQueryRunner_1.PromiseBasedPoolQueryRunner {
    constructor(pool, database = 'mariaDB') {
        super();
        this.pool = pool;
        this.database = database;
    }
    useDatabase(database) {
        if (database !== 'mariaDB' && database !== 'mySql') {
            throw new Error('Unsupported database: ' + database + '. MariaDBQueryRunner only supports mariaDB or mySql databases');
        }
        else {
            // @ts-ignore
            this.database = database;
        }
    }
    getNativeRunner() {
        return this.pool;
    }
    addParam(params, value) {
        params.push(value);
        return '?';
    }
    createQueryRunner() {
        return this.pool.getConnection().then(mariaDBConnection => new MariaDBQueryRunner_1.MariaDBQueryRunner(mariaDBConnection, this.database));
    }
    releaseQueryRunner(queryRunner) {
        queryRunner.getNativeRunner().release();
    }
}
exports.MariaDBPoolQueryRunner = MariaDBPoolQueryRunner;
