// ts-sql-query/queryRunners/Sqlite3WasmOO1QueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sqlite3WasmOO1QueryRunner = void 0;
const SqlTransactionQueryRunner_1 = require("./SqlTransactionQueryRunner");
class Sqlite3WasmOO1QueryRunner extends SqlTransactionQueryRunner_1.SqlTransactionQueryRunner {
    constructor(connection, config) {
        super();
        this.connection = connection;
        this.database = 'sqlite';
        this.promise = config?.promise || Promise;
    }
    useDatabase(database) {
        if (database !== 'sqlite') {
            throw new Error('Unsupported database: ' + database + '. BetterSqlite3QueryRunner only supports sqlite databases');
        }
    }
    getNativeRunner() {
        return this.connection;
    }
    getCurrentNativeTransaction() {
        return undefined;
    }
    execute(fn) {
        return fn(this.connection);
    }
    executeQueryReturning(query, params) {
        try {
            const rows = this.connection.selectObjects(query, params);
            return this.promise.resolve(rows);
        }
        catch (e) {
            return this.promise.reject(e);
        }
    }
    executeMutation(query, params) {
        try {
            this.connection.exec({ sql: query, bind: params });
            return this.promise.resolve(this.connection.changes());
        }
        catch (e) {
            return this.promise.reject(e);
        }
    }
    executeInsertReturningLastInsertedId(query, params = []) {
        if (this.containsInsertReturningClause(query, params)) {
            return super.executeInsertReturningLastInsertedId(query, params);
        }
        try {
            this.connection.exec({ sql: query, bind: params });
            const id = this.connection.selectValue('select last_insert_rowid()');
            return this.promise.resolve(id);
        }
        catch (e) {
            return this.promise.reject(e);
        }
    }
    addParam(params, value) {
        if (typeof value === 'boolean') {
            params.push(Number(value));
        }
        else {
            params.push(value);
        }
        return '?';
    }
    createResolvedPromise(result) {
        return this.promise.resolve(result);
    }
    createAllPromise(promises) {
        return this.promise.all(promises);
    }
}
exports.Sqlite3WasmOO1QueryRunner = Sqlite3WasmOO1QueryRunner;
